osu.Game/Beatmaps/Drawables/Cards/BeatmapCardThumbnail.cs

```diff
@@ -36,14 +35,14 @@ public new MarginPadding Padding
         [Resolved]
         private OverlayColourProvider colourProvider { get; set; } = null!;
 
-        public BeatmapCardThumbnail(APIBeatmapSet beatmapSetInfo)
+        public BeatmapCardThumbnail(IBeatmapSetInfo beatmapSetInfo)
         {
             InternalChildren = new Drawable[]
             {
                 new UpdateableOnlineBeatmapSetCover(BeatmapSetCoverType.List)
                 {
                     RelativeSizeAxes = Axes.Both,
-                    OnlineInfo = beatmapSetInfo
+                    OnlineInfo = beatmapSetInfo as IBeatmapSetOnlineInfo
```

Comment:
I disagree with soft-casting in components like this. Stuff like that leads to half-broken reuse later because someone shoves in here an `IBeatmapInfo` that is not `IBeatmapSetOnlineInfo` and then to compensate various hacks are added and thus we end up with messes like `DrawableAvatar` or whatever which should be entirely online-based but ends up hacking in local hardcoded `a.ppy.sh` URLs to work in all use cases.

`BeatmapCardThumbnail` should accept `IBeatmapSetOnlineInfo` period. No casting inside. The cast can go live in `DrawableRoomPlaylistItem` (e.g. local to the usage, wherein it is much more sane to determine whether the model being passed into drawable components actually *has* all of the data it needs to have).

Reply:
I agree.

The main issue here is that we don't have an interface which defines both `IBeatmapSetOnlineInfo` and `IBeatmapSetInfo`. We'd either have to pass in an `APIBeatmapSet` instead, or create an interface which covers both of those.

Reply:
Maybe pass as two parameters instead, even if the consumer has to pass in `(beatmapSetInfo, beatmapSetInfo)` to the ctor? Don't know how deep this goes.

Reply:
passing as two parameter works, if everyone agrees on this.

Reply:
I'm fine with either 2 params or `IBeatmapSetOnlineInfo : IBeatmapSetInfo` (which should work right now, only two implementors of the former implement both). But passing both is also fine with me.

Reply:
`TournamentBeatmap` is the exception to the rule. I'll see if that's fixable..

Reply:
Is it?

https://github.com/ppy/osu/blob/525663fc27593f8eac6d9f7b71b7f7e5e20f3520/osu.Game/Online/API/Requests/Responses/APIBeatmap.cs#L12
https://github.com/ppy/osu/blob/525663fc27593f8eac6d9f7b71b7f7e5e20f3520/osu.Game.Tournament/Models/TournamentBeatmap.cs#L12

Reply:
Unless I'm missing something, one is combining a beatmap and one is combining a set.

![JetBrains Rider 2024-05-27 at 10 53 45](https://github.com/ppy/osu/assets/191335/0d2f72a5-3cf9-4b65-96d6-2d9e7572736a)

The naive fix would be to implement the interface and throw `NotImplementedException` for the unsupported pieces?

Reply:
This doesn't completely solve the issue unfortunately. There still needs to be a hard or soft cast from `IBeatmapSetInfo` in `DrawableRoomPlaylistItem` (unless we assert that the `BeatmapSet` is an `APIBeatmapSet` or similar):

![JetBrains Rider 2024-05-27 at 10 58 32](https://github.com/ppy/osu/assets/191335/854b569d-b90d-4c9e-8bba-15b58d7bc5ea)

Reply:
> This doesn't completely solve the issue unfortunately. There still needs to be a hard or soft cast from `IBeatmapSetInfo` in `DrawableRoomPlaylistItem`

Yeah I'm fine with that. `DrawableRoomPlaylistItem` is generally very bad code quality wise (does too much, relies on model magic, it needs to be rewritten and split into the 5 actual use cases it attempts to service as one class). So it can keep all of the badness all of the above implies. I just don't want it spreading any further than it already is.